File: cpp/euler3.cpp
#include <iostream>
#include <cmath>

template <typename T>
bool is_prime(T p)
{
	for (T i = 2; i < (p / 2.0); ++i)
		if (fmod(p, i) == 0.0)
			return false;
	return true;
}

template <typename T>
T trial_division_prime_only(T n)
{
	T max = sqrt(n), result = 0.0;
	for (T i = 2; i <= max; ++i)
		if (fmod(n, i) == 0.0 && is_prime(i))
			result = i;
	return result;
}

int main(int argc, char *argv[])
{
	long double n = 600851475143;
	std::cout << "Answer is " << trial_division_prime_only(n) << std::endl;
	return 0;
}

File: lua/euler3.lua
#!/usr/bin/env lua

function isprime(p)
	local n = p / 2
	for i = 2, n do
		if math.fmod(p, i) == 0 then
			return false
		end
	end
	return true
end

function trial_div_prime(n)
	local max, res = math.sqrt(n), 0
	for i = 2, max do
		if math.fmod(n, i) == 0 and isprime(i) then
			res = i
		end
	end
	return res
end

print("Answer is " .. trial_div_prime(600851475143))
